### PR TITLE
Obey ASN.1 substructure bounds in X.509 CRT parsing

### LINKED ISSUE #2431: X.509 CRT parser returns ASN.1 error code on empty TBSCertificate
__Context:__ The first field in a `TBSCertificate` is the version which is parsed through `x509_get_version()`:
https://github.com/ARMmbed/mbedtls/blob/f352f75f6bd5734c8f671323dd6ab32472d5da34/library/x509_crt.c#L380-L400
If this function finds an empty buffer, the call to `mbedtls_asn1_get_tag()` will return `MBEDTLS_ERR_ASN1_OUT_OF_DATA`, which is then returned unmodified to the caller.

__Issue:__ This is not in line with the rest of the module, which wraps low-level ASN.1 error code with high-level X.509 error codes before returning. 

__Task:__ Make sure `x509_crt_parse_der_core()` always returns an X.509 error code. Add a test exercising an empty `TBSCertificate` body.

### LINKED ISSUE #2437: X.509 AttributeTypeAndValue parsing doesn't respect bounds
__Context:__ This is about the parsing of the X.509 structure
```
   AttributeTypeAndValue ::= SEQUENCE {
     type     AttributeType,
     value    AttributeValue }
```
in `x509_get_attr_type_and_value()` from `library/x509.c`:
https://github.com/ARMmbed/mbedtls/blob/f352f75f6bd5734c8f671323dd6ab32472d5da34/library/x509.c#L344-L370

__Issue:__ The function doesn't obey the `len` bound for the outer `SEQUENCE`. Instead, it only makes sure that the parsing doesn't read past the surrounding 
```
   RelativeDistinguishedName ::=
     SET SIZE (1..MAX) OF AttributeTypeAndValue
```
the end of which is passed to the function as `end`. For example, if the length of the outer `SEQUENCE` is `0` but there's more space remaining in the current `SET`, the function won't fail. This situation is actually exercised by the test https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_x509parse.data#L1001 which wrongly expects `MBEDTLS_ERR_X509_INVALID_NAME + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG`, where it should be `MBEDTLS_ERR_X509_INVALID_NAME + MBEDTLS_ERR_ASN1_OUT_OF_DATA`.

This is not a security issue because the bounds of the surrounding `SET` are obeyed.

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR fixes #2431 and fixes #2437. See there for more information on the underlying problem, and the commit messages in this PR for more information on the approach taken.

The PR has conflicts with #2430 and is based on the latter to resolve them.

### CODE REVIEW

File: library/x509.c [hunk 1]
```diff
@@ -354,6 +354,8 @@ static int x509_get_attr_type_value( unsigned char **p,
             MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
         return( MBEDTLS_ERR_X509_INVALID_NAME + ret );
 
+    end = *p + len;
```
Comment (Patater) [inline library/x509.c]: Does this make the below assignment of `oid->tag = **p;` wrong? since before we had `oid->tag` pointing to the beginning of the tag. Now, we have it pointing to the end of the tag.

Consider setting `oid->tag` to `MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE` explicitly.

File: library/x509.c [hunk 2]
```diff
@@ -693,30 +701,25 @@ int mbedtls_x509_get_sig_alg( const mbedtls_x509_buf *sig_oid, const mbedtls_x50
...
+    ext->p   = *p;
+    end      = *p + ext->len;
 
     /*
      * Extensions  ::=  SEQUENCE SIZE (1..MAX) OF Extension
-     *
-     * Extension  ::=  SEQUENCE  {
-     *      extnID      OBJECT IDENTIFIER,
-     *      critical    BOOLEAN DEFAULT FALSE,
-     *      extnValue   OCTET STRING  }
```
Comment (Patater) [inline library/x509.c]: Is it intentional to remove this comment explaining what "Extension" is? I read the commend as explaining what an Extension was, not that this comment was saying we parsed Extension here. Should we move the Extension documentation elsewhere or does it exist somewhere else already?

File: library/x509_crt.c [hunk 3]
```diff
@@ -393,7 +393,7 @@ static int x509_get_version( unsigned char **p,
             return( 0 );
         }
 
-        return( ret );
+        return( MBEDTLS_ERR_X509_INVALID_FORMAT + ret );
```
Comment (Patater) [inline library/x509_crt.c]: Is it guaranteed this won't ever overflow? Do we want to use `+` or `|`?

File: library/x509.c (hunk 1 above)
Comment (hanno-arm) [reply library/x509.c]: I don't follow - the line modifies `end`, but not `*p`, hence not `oid->tag`. Am I missing something?

Comment (hanno-arm) [reply library/x509.c]: `oid->tag` will necessarily be `MBEDTLS_ASN1_OID` in the end, and one could move `oid->tag = **p` to `oid->tag = MBEDTLS_ASN1_OID` once the `mbedtls_asn1_get_tag()` call succeeded. Maybe the latter saves a few bytes, so let's do it.

File: library/x509.c (hunk 2 above)
Comment (hanno-arm) [reply library/x509.c]: Yes, it's fully intentional - the comment is very misleading, because it suggests that we're parsing an `Extension` structure (or its header) here, which is _not_ the case. We're parsing the two (!) headers of the `Extensions` structure: the explicit tag, followed by the `SEQUENCE OF` tag. Afterwards, `*p` points to the beginning of the header of the `Extension` structure, but the shape of the latter is entirely irrelevant to this function -- at least if I understand the code correctly.

File: library/x509_crt.c (hunk 3 above)
Comment (hanno-arm) [reply library/x509_crt.c]: Yes, we're returning an ASN.1 low level error code here, and wrapping it by a high-level X.509 error code. This is the idiom that we have throughout the library. As mentioned elsewhere, it's not optimal, because what _we_ know the compiler doesn't, namely that the error codes of X.509 and ASN.1 are binary disjoint; using `|` instead of `+` would at least allow the compiler or static analyzers to realize that the wrapping doesn't lead to zero. However, as mentioned, it's a common idiom throughout the library, so it should be changed everywhere or nowhere, but not in a single place.

File: library/x509.c (hunk 1 above)
Comment (hanno-arm) [reply library/x509.c]: Would you be OK with leaving this change to a separate X.509 PR (there are enough waiting), so that we don't undo @k-stachowiak's review in case there's nothing else to be changed?

File: library/x509.c (hunk 2 above)
Comment (hanno-arm) [reply library/x509.c]: The comment is duplicated correctly in https://github.com/ARMmbed/mbedtls/blob/908cb42d7d3eb4322f2aa149ad5201ab7f6828c2/library/x509_crt.c#L711-L715 where it belongs.

File: library/x509_crt.c (hunk 3 above)
Comment (Patater) [reply library/x509_crt.c]: OK. Not that happy we are justifying not doing things the best we can in new code because we didn't in old code, but fine.

File: library/x509.c (hunk 1 above)
Comment (Patater) [reply library/x509.c]: Yup, I'm misreading the code. `p` hasn't moved.

File: library/x509.c (hunk 2 above)
Comment (Patater) [reply library/x509.c]: OK

File: library/x509_crt.c (hunk 3 above)
Comment (hanno-arm) [reply library/x509_crt.c]: I think it's better to know we're always _consistently_ using a particular idiom, even if it's not optimal and could be improved, than to mix idioms and loose track of where we already switched and where we didn't. Don't you think? I created https://github.com/ARMmbed/mbedtls/issues/2462.

**Comment by hanno-arm** [comment]
@Patater Could you potentially switch the state to "Ready for Merge" and add this PR to the list of candidates for the next merge to `development`? Thanks!

**Comment by hanno-arm** [comment]
@Patater @k-stachowiak Can you please re-review ths latest rebase on top of the updated #2430?

File: ChangeLog [hunk 4]
```diff
@@ -130,6 +130,11 @@ Bugfix
      extensions in CSRs and CRTs that caused these bitstrings to not be encoded
      correctly as trailing zeroes were not accounted for as unused bits in the
      leading content octet. Fixes #1610.
+   * Fix missing bounds checks in X.509 parsing functions that could
+     lead to successful parsing of ill-formed X.509 CRTs. Fixes #2437.
+   * Fix multiple X.509 functions previously returning ASN.1 low-level error
+     codes to always wrap these codes into X.509 high level error codes before
+     returning. Fixes #2431.
```
Comment (Patater) [inline ChangeLog]: This is being added to 2.17.0 instead of the latest release. Will fix in merging.

File: ChangeLog [hunk 5]
```diff
@@ -49,6 +49,11 @@ Bugfix
    * Set the next sequence of the subject_alt_name to NULL when deleting
      sequence on failure. Found and fix suggested by Philippe Antoine.
      Credit to OSS-Fuzz.
+   * Fix missing bounds checks in X.509 parsing functions that could
```
Comment (Patater) [inline ChangeLog]: Oh, nevermind. You've added another commit to move it.

**Comment by Patater** [comment]
https://github.com/ARMmbed/mbedtls/pull/2430 has been merged; removing label "needs: preceding PR"
